### ISSUE TITLE
Why is (SingKind a) needed for phantom type?

### ISSUE THREAD

**Issue by LeanderK** [goldfirere/singletons#348]
I have:

```haskell
$(singletons [d|
data ExplBasisG (a :: Type) (b :: Type) = ExplBasisG [b]
        deriving (Show, Eq)
  |])
```

I would expect to only need a `SingKind b` for a `fromSing`, since a is a type and irrelevant for fromSing.

**Comment by RyanGlScott** [comment]
I've grappled with this question before. The question one must ask is: what should the promoted version of `ExplBasisG (a -> b) c` be? Should it be:

* `ExplBasisG (a -> b) c`?
* `ExplBasisG (a ~> b) c`?

If we pick the former, then we needn't have a `SingKind a` constraint in the instance for `ExplBasisG`. But it does mean that it's less expressive than it might otherwise be, because fewer types inhabit the kind `a -> b` than `a ~> b`.

If we pick the latter, then the promoted version is as expressive at the type level as possible, but it does require that we impose a `SingKind a` constraint.

Similar questions arise if you decide to ask what the promoted versions of `ExplBasisG Natural b` and `ExplBasisG Text b` should be.

I originally noticed this back when I realized that other derived instances were [imposing redundant constraints](https://github.com/goldfirere/singletons/issues/271) on phantom type variables, but because I couldn't come up with a satisfactory answer to the question above, I left `SingKind`'s instance context inference alone.
